tests/devices/qubit/test_apply_operation.py

```diff
@@ -1267,3 +1274,118 @@ def circuit(init_state):
         assert qml.math.shape(results) == (3, 256)
         assert np.array_equal(results[:, 128], [-1.0 + 0.0j] * 3)
+
+
+# pylint: disable=too-few-public-methods
+class TestConditionalsAndMidMeasure:
+    """Test dispatching for mid-circuit measurements and conditionals."""
+
+    @pytest.mark.parametrize("ml_framework", ml_frameworks_list)
+    @pytest.mark.parametrize("batched", (False, True))
+    @pytest.mark.parametrize("unitary", (qml.CRX, qml.CRZ))
+    @pytest.mark.parametrize("wires", ([0, 1], [1, 0]))
+    def test_conditional(self, wires, unitary, batched, ml_framework):
+        """Test the application of a Conditional on an arbitrary state."""
+
+        n_states = int(batched) + 1
+        initial_state = np.array(
+            [
+                [
+                    0.3541035 + 0.05231577j,
+                    0.6912382 + 0.49474503j,
+                    0.29276263 + 0.06231887j,
+                    0.10736635 + 0.21947607j,
+                ],
+                [
+                    0.09803567 + 0.47557068j,
+                    0.4427561 + 0.13810454j,
+                    0.26421703 + 0.5366283j,
+                    0.03825933 + 0.4357423j,
+                ],
+            ][:n_states]
+        )
+
+        rotated_state = qml.math.dot(
+            initial_state, qml.matrix(unitary(-0.238, wires), wire_order=[0, 1]).T
+        )
+        rotated_state = qml.math.asarray(rotated_state, like=ml_framework)
+        rotated_state = qml.math.squeeze(qml.math.reshape(rotated_state, (n_states, 2, 2)))
+
+        m0 = qml.measure(0)
+        op = qml.ops.op_math.Conditional(m0, unitary(0.238, wires))
+
+        mid_meas = {m0.measurements[0]: 0}
+        old_state = apply_operation(
+            op, rotated_state, batched, interface=ml_framework, mid_measurements=mid_meas
+        )
+        assert qml.math.allclose(rotated_state, old_state)
+
+        mid_meas[m0.measurements[0]] = 1
+        new_state = apply_operation(
+            op, rotated_state, batched, interface=ml_framework, mid_measurements=mid_meas
+        )
+        assert qml.math.allclose(
+            qml.math.squeeze(initial_state), qml.math.reshape(new_state, (n_states, 4))
+        )
+
+    @pytest.mark.parametrize("rng_seed, m_res", ((12, (0, 0)), (42, (1, 1))))
+    def test_mid_measure(self, rng_seed, m_res):
+        """Test the application of a MidMeasureMP on an arbitrary state to give a basis state."""
+
+        initial_state = np.array(
+            [
+                [0.09068964 + 0.36775595j, 0.37578343 + 0.4786927j],
+                [0.3537292 + 0.27214766j, 0.01928256 + 0.53536021j],
+            ]
+        )
+
+        mid_state, end_state = np.zeros((2, 2), dtype=complex), np.zeros((2, 2), dtype=complex)
+        mid_state[m_res[0]] = initial_state[m_res[0]] / np.linalg.norm(initial_state[m_res[0]])
+        end_state[m_res] = mid_state[m_res] / np.abs(mid_state[m_res])
+
+        rng = np.random.default_rng(rng_seed)
```

Comment:
Why is seeding necessary here? Since the mid measure results are manually set by the dictionary, there shouldn't be any randomness to what the result state is?

Reply:
As far as I could understand: the mid-measure results here are not set by the `dictionary` and are rather set here:
https://github.com/PennyLaneAI/pennylane/blob/6903275a31a20d1c3700f6f3deb9a96c3fdc3d95/pennylane/devices/qubit/apply_operation.py#L343
based on the `binomial` distribution. I am setting the `seed` to fix the results from the sample.

Reply:
In that case, what does setting the dictionary do?

Reply:
It looks to me that the dictionary here is just for bookkeeping purposes and simply ensures `mid_measurements[op]` doesn't raise a `KeyError` in this scenario. 😅

Reply:
The dictionary is just for storing the outcomes. The actual measured value is determined non-deterministically by `apply_operation`

Reply:
In that case why do we set
```
mid_meas = {m0: m_res[0], m1: m_res[1]}
```
below? What does this do?

Reply:
This is a remnant from the previous test case, this just ensures that `m0` and `m1` keys are available when we do `mid_measurement[m0] = sample` in `_apply_operation` (otherwise it will throw a `KeyError`). So, in reality we could also do `mid_meas = {m0: None, m1: None}`

Reply:
> this just ensures that m0 and m1 keys are available when we do mid_measurement[m0] = sample in _apply_operation (otherwise it will throw a KeyError)

That shouldn't be the case. `KeyError`s are raised when you try to get a value for a key that doesn't exist. The example you used is _setting_ the value for a key, which works even if the key doesn't exist in the dictionary. In that case it just adds that key to the dictionary.

Reply:
Checked again, you're right. It works here, maybe it was somewhere else. 👍